cmd/egress-v6-cni-plugin/cni.go

```diff
@@ -0,0 +1,608 @@
+	if hostPrimaryIfIPv6 == nil || !hostPrimaryIfIPv6.IsGlobalUnicast() {
+		return fmt.Errorf("gobal unicast IPv6 address is needed on host primary interface for IPv6 egress")
+	}
+	log.Debugf("IPv6 address retrieved: %s", hostPrimaryIfIPv6AddrStr)
+
+	chain := utils.MustFormatChainNameWithPrefix(netConf.Name, args.ContainerID, "E6-")
+	comment := utils.FormatComment(netConf.Name, args.ContainerID)
+
+	netns, err := ns.GetNS(args.Netns)
+	if err != nil {
+		log.Debugf("failed to open netns %q: %v", args.Netns, err)
+		return fmt.Errorf("failed to open netns %q: %v", args.Netns, err)
+	}
+	defer netns.Close()
+
+	// assign a ULA (Unique Local Address) ipv6 address in POD
+	containerIPNetV6, ipamResult, err := setupContainerIPv6Address(result, netConf.IPAM.Type, netns, args.IfName, args.StdinData)
+	if err != nil {
+		log.Debugf("setupContainerIPv6Address failed: %v", err)
+		return err
+	}
+
+	log.Debugf("container IPv6 assigned: %s", containerIPNetV6.String())
+
+	err = setupContainerIPv6Route(netns, containerIPNetV6, args.IfName)
+	if err != nil {
+		log.Debugf("setupContainerIPv6Route failed: %v", err)
+		return err
+	}
+	log.Debugf("container route set up successfully")
+
+	err = enableHostIPv6Forwarding()
+	if err != nil {
+		log.Debugf("enableHostIPv6Forwarding failed: %v", err)
+		return err
+	}
+	log.Debugf("enable host IPv6 forwarding successfully")
+
+	err = setupHostIPv6Route(netns, args.IfName, containerIPNetV6.IP)
+	if err != nil {
+		log.Debugf("setupHostIPv6Route failed: %v", err)
+		return err
+	}
+	log.Debugf("host IPv6 route set up successfully")
+
+	err = setupHostIPv6Snat(containerIPNetV6.IP, hostPrimaryIfIPv6, chain, comment, netConf.RandomizeSNAT)
+	if err != nil {
+		log.Debugf("setupHostIPv6Snat failed: %v", err)
+		return err
+	}
+
+	log.Debugf("host IPv6 SNAT set up successfully")
+
+	// Copy IPs over to result
+	result.IPs = append(result.IPs, ipamResult.IPs...)
+	// Note: Useful for debug, will do away with the below log prior to release
+	for _, v := range result.IPs {
+		log.Debugf("Interface index: %d; IP: %s", *v.Interface, v.Address)
+	}
+
+	// Pass through the previous result
+	return types.PrintResult(result, netConf.CNIVersion)
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+
+	netConf, log, err := loadConf(args.StdinData)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	log.Debugf("Received Del request for: conf=%v; Plugin enabled=%s", netConf, netConf.Enabled)
+
+	// We only need this plugin to kick in if v6 is enabled
+	if netConf.Enabled == "false" {
+		return nil
+	}
+
+	if netConf.PrevResult == nil {
+		log.Debugf("must be called as a chained plugin")
+		return fmt.Errorf("must be called as a chained plugin")
+	}
+
+	result, err := current.GetResult(netConf.PrevResult)
+	if err != nil {
+		log.Debugf("get PrevResult failed: %v", err)
+		return err
+	}
+	log.Debugf("PrevResult: %v", result)
+
+	// figure out container IPv6 address
```

Comment:
So at this point, the main VPC CNI plugin has already run, but since it does not delete the `eth0` interface, we can still derive the IPv6 ULA address that we assigned to it. Thinking out loud, the only concern here is if the main VPC CNI plugin changes its functionality someday to delete the `eth0` interface. We can discuss more offline about how we can protect from this regression ever happening.